google/api_core/retry_streaming_async.py

```diff
@@ -0,0 +1,266 @@
+        continue iterating. If not, raise the exception.
+        """
+        if not self.predicate(exc) and not isinstance(exc, asyncio.TimeoutError):
+            raise exc
+        else:
+            # run on_error callback if provided
+            if self.on_error:
+                self.on_error(exc)
+            try:
+                next_sleep = next(self.sleep_generator)
+            except StopIteration:
+                raise ValueError("Sleep generator stopped yielding sleep values")
+            # if time budget is exceeded, raise RetryError
+            if self.remaining_timeout_budget is not None and self.timeout is not None:
+                if self.remaining_timeout_budget <= next_sleep:
+                    raise exceptions.RetryError(
+                        "Timeout of {:.1f}s exceeded".format(self.timeout),
+                        exc,
+                    ) from exc
+                else:
+                    self.remaining_timeout_budget -= next_sleep
+            _LOGGER.debug(
+                "Retrying due to {}, sleeping {:.1f}s ...".format(exc, next_sleep)
+            )
+            # sleep before retrying
+            await asyncio.sleep(next_sleep)
+            self.active_target = None
+            await self._ensure_active_target()
+
+    def _subtract_time_from_budget(self, start_timestamp: datetime.datetime) -> None:
+        """
+        Subtract the time elapsed since start_timestamp from the remaining
+        timeout budget.
+
+        Args:
+        - start_timestamp (datetime): The time at which the last operation
+            started.
+        """
+        if self.remaining_timeout_budget is not None:
+            self.remaining_timeout_budget -= (
+                datetime_helpers.utcnow() - start_timestamp
+            ).total_seconds()
+
+    async def _iteration_helper(self, iteration_routine: Awaitable) -> T:
+        """
+        Helper function for sharing logic between __anext__ and asend.
+
+        Args:
+          - iteration_routine: The coroutine to await to get the next value
+                from the iterator (e.g. __anext__ or asend)
+        Returns:
+          - The next value from the active_target iterator.
+        """
+        # check for expired timeouts before attempting to iterate
+        if (
+            self.remaining_timeout_budget is not None
+            and self.remaining_timeout_budget <= 0
+            and self.timeout is not None
+        ):
+            raise exceptions.RetryError(
+                "Timeout of {:.1f}s exceeded".format(self.timeout),
+                None,
+            )
+        try:
+            # start the timer for the current operation
+            start_timestamp = datetime_helpers.utcnow()
+            # grab the next value from the active_target
+            next_val_routine = asyncio.wait_for(
+                iteration_routine, self.remaining_timeout_budget
+            )
```

Comment:
After doing some bench marking, it looks like `wait_for` slows things down significantly.

I'm thinking it might be best to have this follow the approach used by the synchronous version, and only timeout after the await is finished, rather than interrupting it early

Alternatively, Python3.11 has new timeout features which I believe are more performant, (along with a [third-party library](https://github.com/aio-libs/async-timeout) that is backwards compatible). But probably best to keep things simple.

Let me know what you think